File: 22FALL/79.word-search.py
#
# @lc app=leetcode id=79 lang=python3
#
# [79] Word Search
#

# @lc code=start
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        moves = [0, 1, 0, -1, 0]
        
        def dfs(i, j, c_i):            
            if c_i == len(word):
                return True
            if i >= m or j >=n or i < 0 or j<0 or board[i][j] != word[c_i]:
                return False
            tmp = board[i][j]
            board[i][j] = '0'
            for k in range(4):
                dx, dy = i + moves[k], j + moves[k+1]
                if dfs(dx, dy, c_i+1):
                    return True
            board[i][j] = tmp
            return False
        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True
        return False

# @lc code=end

File: VSCode_CPP/79.word-search.cpp
/*
 * @lc app=leetcode id=79 lang=cpp
 *
 * [79] Word Search
 */

// @lc code=start
class Solution {
   public:
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size(), n = board[0].size();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, word, i, j, 0)) return true;
            }
        }
        return false;
    }

   private:
    int m, n;
    bool dfs(vector<vector<char>>& board, string word, int i, int j, int k) {
        if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != word[k])
            return false;
        if (k == word.size() - 1) return true;
        board[i][j] = '\0';
        bool dfs_rs = dfs(board, word, i + 1, j, k + 1) ||
                      dfs(board, word, i - 1, j, k + 1) ||
                      dfs(board, word, i, j + 1, k + 1) ||
                      dfs(board, word, i, j - 1, k + 1);
        board[i][j] = word[k];
        return dfs_rs;
    }
};
// @lc code=end
